src/slave/containerizer/mesos/isolators/network/port_mapping.hpp

```diff
@@ -152,6 +153,59 @@ Result<routing::queueing::htb::cls::Config> recoverHTBConfig(
     const std::string& eth0,
     const Flags& flags);
 
+
+class PercentileRatesCollector final
+{
+public:
+  explicit PercentileRatesCollector(
+      pid_t _executorPid, const Duration& _window, const Duration& _interval);
+
+  Option<process::Statistics<uint64_t>> rxRate() const;
+  Option<process::Statistics<uint64_t>> rxPacketRate() const;
+  Option<process::Statistics<uint64_t>> rxDropRate() const;
+  Option<process::Statistics<uint64_t>> rxErrorRate() const;
+
+  Option<process::Statistics<uint64_t>> txRate() const;
+  Option<process::Statistics<uint64_t>> txPacketRate() const;
+  Option<process::Statistics<uint64_t>> txDropRate() const;
+  Option<process::Statistics<uint64_t>> txErrorRate() const;
+
+  // Sample statistics from the interface.
+  void sample();
+
+  // Register the statistics sample. Exposed for testing.
+  void sample(const process::Time& ts, hashmap<std::string, uint64_t>&& stats);
+
+private:
+  void sampleRate(
+      const hashmap<std::string, uint64_t>& statistics,
+      const std::string& metric,
+      const process::Time& timestamp,
+      double timeDelta,
+      process::TimeSeries<uint64_t>& rates);
+
+  // Name of the link to sample metrics from.
+  const std::string link;
+
+  process::TimeSeries<uint64_t> rxRates;
+  process::TimeSeries<uint64_t> rxPackets;
+  process::TimeSeries<uint64_t> rxDrops;
+  process::TimeSeries<uint64_t> rxErrors;
+
+  process::TimeSeries<uint64_t> txRates;
+  process::TimeSeries<uint64_t> txPackets;
+  process::TimeSeries<uint64_t> txDrops;
+  process::TimeSeries<uint64_t> txErrors;
+
+  // Previous sample and its timestamp for calculating rates.
+  Option<process::Time> previous;
+  hashmap<std::string, uint64_t> previousStatistics;
```

Comment:
Some follow-on optimizations: since in practice we're sampling every 50ms, it's probably worth avoiding the string keyed map here and instead use an enum indexed array like the library underneath us uses, which is significantly more efficient.